File: Data Structures & Algorithms/balanced-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private boolean balanced = true;
    public boolean isBalanced(TreeNode root) {
        helper(root);
        return balanced;
    }

    private int helper(TreeNode node) {
        if (node == null || balanced == false)
            return 0;

        int left = helper(node.left);
        int right = helper(node.right);

        if (Math.abs(left - right) > 1)
            balanced = false;

        return Math.max(left, right) + 1;
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    res = [0]
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root: TreeNode | null): number {
        
        this.helper(root)
        return this.res[0]
    }

    helper(root: TreeNode | null): number {
        if (root === null) return 0

        let left = this.helper(root.left)
        let right = this.helper(root.right)

        this.res[0] = Math.max(left + right, this.res[0])
        return 1 + Math.max(left, right)
    }
}
